src/Core/IdentityServer/ResourceOwnerPasswordValidator.cs

```diff
@@ -135,5 +134,33 @@ protected async override Task<(User, bool)> ValidateContextAsync(ResourceOwnerPa
         {
             context.Result = new GrantValidationResult(TokenRequestErrors.InvalidGrant, customResponse: customResponse);
         }
+
+        private bool AuthEmailHeaderIsValid(ResourceOwnerPasswordValidationContext context)
+        {
+            if (!_currentContext.HttpContext.Request.Headers.ContainsKey("Auth-Email"))
+            {
+                return false;
+            }
+            else
+            {
+                try
+                {
+                    var authEmailHeader = _currentContext.HttpContext.Request.Headers["Auth-Email"];
+                    var authEmailDecoded = CoreHelpers.Base64UrlDecodeString(authEmailHeader);
+
+                    if (authEmailDecoded != context.UserName)
+                    {
+                        return false;
+                    }
+                }
+                catch (System.Exception e) when (e is System.InvalidOperationException || e is System.FormatException)
+                {
+                    // Invalid B64 encoding
+                    return false;
+                }
+            }
+
+            return true;
```

Comment:
This could then be removed if using the prior suggestion, further simplifying the flow.